File: jscomp/test/sexp.js
'use strict';

var List = require("../../lib/js/list.js");
var Curry = require("../../lib/js/curry.js");
var Hashtbl = require("../../lib/js/hashtbl.js");
var Caml_obj = require("../../lib/js/caml_obj.js");
var Pervasives = require("../../lib/js/pervasives.js");
var Caml_format = require("../../lib/js/caml_format.js");
var Caml_option = require("../../lib/js/caml_option.js");

var equal = Caml_obj.caml_equal;

var compare = Caml_obj.caml_compare;

var hash = Hashtbl.hash;

function of_int(x) {
  return {
          HASH: /* Atom */726615281,
          value: String(x)
        };
}

function of_float(x) {
  return {
          HASH: /* Atom */726615281,
          value: Pervasives.string_of_float(x)
        };
}

function of_bool(x) {
  return {
          HASH: /* Atom */726615281,
          value: x ? "true" : "false"
        };
}

function atom(x) {
  return {
          HASH: /* Atom */726615281,
          value: x
        };
}

function of_list(l) {
  return {
          HASH: /* List */848054398,
          value: l
        };
}

function of_rev_list(l) {
  return {
          HASH: /* List */848054398,
          value: List.rev(l)
        };
}

function of_pair(param) {
  return {
          HASH: /* List */848054398,
          value: /* :: */{
            _0: param[0],
            _1: /* :: */{
              _0: param[1],
              _1: /* [] */0
            }
          }
        };
}

function of_triple(param) {
  return {
          HASH: /* List */848054398,
          value: /* :: */{
            _0: param[0],
            _1: /* :: */{
              _0: param[1],
              _1: /* :: */{
                _0: param[2],
                _1: /* [] */0
              }
            }
          }
        };
}

function of_quad(param) {
  return {
          HASH: /* List */848054398,
          value: /* :: */{
            _0: param[0],
            _1: /* :: */{
              _0: param[1],
              _1: /* :: */{
                _0: param[2],
                _1: /* :: */{
                  _0: param[3],
                  _1: /* [] */0
                }
              }
            }
          }
        };
}

function of_variant(name, args) {
  return {
          HASH: /* List */848054398,
          value: /* :: */{
            _0: {
              HASH: /* Atom */726615281,
              value: name
            },
            _1: args
          }
        };
}

function of_field(name, t) {
  return {
          HASH: /* List */848054398,
          value: /* :: */{
            _0: {
              HASH: /* Atom */726615281,
              value: name
            },
            _1: /* :: */{
              _0: t,
              _1: /* [] */0
            }
          }
        };
}

function of_record(l) {
  return {
          HASH: /* List */848054398,
          value: List.map((function (param) {
                  return of_field(param[0], param[1]);
                }), l)
        };
}

function $$return(x) {
  return Caml_option.some(x);
}

function $great$pipe$eq(e, f) {
  if (e !== undefined) {
    return Caml_option.some(Curry._1(f, Caml_option.valFromOption(e)));
  }
  
}

function $great$great$eq(e, f) {
  if (e !== undefined) {
    return Curry._1(f, Caml_option.valFromOption(e));
  }
  
}

function map_opt(f, l) {
  var _acc = /* [] */0;
  var _l = l;
  while(true) {
    var l$1 = _l;
    var acc = _acc;
    if (!l$1) {
      return List.rev(acc);
    }
    var y = Curry._1(f, l$1._0);
    if (y === undefined) {
      return ;
    }
    _l = l$1._1;
    _acc = /* :: */{
      _0: Caml_option.valFromOption(y),
      _1: acc
    };
    continue ;
  };
}

function list_any(f, e) {
  if (e.HASH >= 848054398) {
    var _l = e.value;
    while(true) {
      var l = _l;
      if (!l) {
        return ;
      }
      var res = Curry._1(f, l._0);
      if (res !== undefined) {
        return res;
      }
      _l = l._1;
      continue ;
    };
  }
  
}

function list_all(f, e) {
  if (e.HASH >= 848054398) {
    var _acc = /* [] */0;
    var _l = e.value;
    while(true) {
      var l = _l;
      var acc = _acc;
      if (!l) {
        return List.rev(acc);
      }
      var tl = l._1;
      var y = Curry._1(f, l._0);
      if (y !== undefined) {
        _l = tl;
        _acc = /* :: */{
          _0: Caml_option.valFromOption(y),
          _1: acc
        };
        continue ;
      }
      _l = tl;
      continue ;
    };
  } else {
    return /* [] */0;
  }
}

function _try_atom(e, f) {
  if (e.HASH >= 848054398) {
    return ;
  }
  try {
    return Caml_option.some(Curry._1(f, e.value));
  }
  catch (exn){
    return ;
  }
}

function to_int(e) {
  return _try_atom(e, Caml_format.caml_int_of_string);
}

function to_bool(e) {
  return _try_atom(e, Pervasives.bool_of_string);
}

function to_float(e) {
  return _try_atom(e, Caml_format.caml_float_of_string);
}

function to_string(e) {
  return _try_atom(e, (function (x) {
                return x;
              }));
}

function to_pair(e) {
  if (typeof e === "number") {
    return ;
  }
  if (e.HASH !== 848054398) {
    return ;
  }
  var match = e.value;
  if (!match) {
    return ;
  }
  var match$1 = match._1;
  if (match$1 && !match$1._1) {
    return /* tuple */[
            match._0,
            match$1._0
          ];
  }
  
}

function to_pair_with(f1, f2, e) {
  return $great$great$eq(to_pair(e), (function (param) {
                var y = param[1];
                return $great$great$eq(Curry._1(f1, param[0]), (function (x) {
                              return $great$great$eq(Curry._1(f2, y), (function (y) {
                                            return /* tuple */[
                                                    x,
                                                    y
                                                  ];
                                          }));
                            }));
              }));
}

function to_triple(e) {
  if (typeof e === "number") {
    return ;
  }
  if (e.HASH !== 848054398) {
    return ;
  }
  var match = e.value;
  if (!match) {
    return ;
  }
  var match$1 = match._1;
  if (!match$1) {
    return ;
  }
  var match$2 = match$1._1;
  if (match$2 && !match$2._1) {
    return /* tuple */[
            match._0,
            match$1._0,
            match$2._0
          ];
  }
  
}

function to_triple_with(f1, f2, f3, e) {
  return $great$great$eq(to_triple(e), (function (param) {
                var z = param[2];
                var y = param[1];
                return $great$great$eq(Curry._1(f1, param[0]), (function (x) {
                              return $great$great$eq(Curry._1(f2, y), (function (y) {
                                            return $great$great$eq(Curry._1(f3, z), (function (z) {
                                                          return /* tuple */[
                                                                  x,
                                                                  y,
                                                                  z
                                                                ];
                                                        }));
                                          }));
                            }));
              }));
}

function to_list(e) {
  if (e.HASH >= 848054398) {
    return Caml_option.some(e.value);
  }
  
}

function to_list_with(f, e) {
  if (e.HASH >= 848054398) {
    return map_opt(f, e.value);
  }
  
}

function get_field(name, e) {
  if (e.HASH >= 848054398) {
    var _l = e.value;
    while(true) {
      var l = _l;
      if (!l) {
        return ;
      }
      var match = l._0;
      if (typeof match === "number") {
        _l = l._1;
        continue ;
      }
      if (match.HASH !== 848054398) {
        _l = l._1;
        continue ;
      }
      var match$1 = match.value;
      if (match$1) {
        var match$2 = match$1._0;
        if (typeof match$2 === "number") {
          _l = l._1;
          continue ;
        }
        if (match$2.HASH !== 726615281) {
          _l = l._1;
          continue ;
        }
        var match$3 = match$1._1;
        if (match$3) {
          if (match$3._1) {
            _l = l._1;
            continue ;
          }
          if (Caml_obj.caml_equal(name, match$2.value)) {
            return match$3._0;
          }
          _l = l._1;
          continue ;
        }
        _l = l._1;
        continue ;
      }
      _l = l._1;
      continue ;
    };
  }
  
}

function field(name, f, e) {
  return $great$great$eq(get_field(name, e), f);
}

function _get_field_list(name, _l) {
  while(true) {
    var l = _l;
    if (!l) {
      return ;
    }
    var match = l._0;
    if (typeof match === "number") {
      _l = l._1;
      continue ;
    }
    if (match.HASH !== 848054398) {
      _l = l._1;
      continue ;
    }
    var match$1 = match.value;
    if (match$1) {
      var match$2 = match$1._0;
      if (typeof match$2 === "number") {
        _l = l._1;
        continue ;
      }
      if (match$2.HASH !== 726615281) {
        _l = l._1;
        continue ;
      }
      if (Caml_obj.caml_equal(name, match$2.value)) {
        return match$1._1;
      }
      _l = l._1;
      continue ;
    }
    _l = l._1;
    continue ;
  };
}

function field_list(name, f, e) {
  if (e.HASH >= 848054398) {
    return $great$great$eq(_get_field_list(name, e.value), f);
  }
  
}

function _get_variant(s, args, _l) {
  while(true) {
    var l = _l;
    if (!l) {
      return ;
    }
    var match = l._0;
    if (Caml_obj.caml_equal(s, match[0])) {
      return Curry._1(match[1], args);
    }
    _l = l._1;
    continue ;
  };
}

function get_variant(l, e) {
  if (e.HASH < 848054398) {
    return _get_variant(e.value, /* [] */0, l);
  }
  var match = e.value;
  if (!match) {
    return ;
  }
  var match$1 = match._0;
  if (typeof match$1 === "number" || match$1.HASH !== 726615281) {
    return ;
  } else {
    return _get_variant(match$1.value, match._1, l);
  }
}

function get_exn(e) {
  if (e !== undefined) {
    return Caml_option.valFromOption(e);
  }
  throw {
        RE_EXN_ID: "Failure",
        _1: "CCSexp.Traverse.get_exn",
        Error: new Error()
      };
}

var of_unit = {
  HASH: /* List */848054398,
  value: /* [] */0
};

var Traverse = {
  map_opt: map_opt,
  list_any: list_any,
  list_all: list_all,
  to_int: to_int,
  to_string: to_string,
  to_bool: to_bool,
  to_float: to_float,
  to_list: to_list,
  to_list_with: to_list_with,
  to_pair: to_pair,
  to_pair_with: to_pair_with,
  to_triple: to_triple,
  to_triple_with: to_triple_with,
  get_field: get_field,
  field: field,
  get_variant: get_variant,
  field_list: field_list,
  $great$great$eq: $great$great$eq,
  $great$pipe$eq: $great$pipe$eq,
  $$return: $$return,
  get_exn: get_exn
};

exports.equal = equal;
exports.compare = compare;
exports.hash = hash;
exports.atom = atom;
exports.of_int = of_int;
exports.of_bool = of_bool;
exports.of_list = of_list;
exports.of_rev_list = of_rev_list;
exports.of_float = of_float;
exports.of_unit = of_unit;
exports.of_pair = of_pair;
exports.of_triple = of_triple;
exports.of_quad = of_quad;
exports.of_variant = of_variant;
exports.of_field = of_field;
exports.of_record = of_record;
exports.Traverse = Traverse;
/* No side effect */

File: jscomp/test/sexp.ml
(* Copyright (c) 2013, Simon Cruanes All rights reserved.

   Redistribution and use in source and binary forms, with or without
   modification, are permitted provided that the following conditions are met:

   Redistributions of source code must retain the above copyright notice, this
   list of conditions and the following disclaimer. Redistributions in binary
   form must reproduce the above copyright notice, this list of conditions and
   the following disclaimer in the documentation and/or other materials
   provided with the distribution.

   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
   AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
   IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
   ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
   LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
   CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
   SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
   INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
   CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
   ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
   POSSIBILITY OF SUCH DAMAGE. *)

(** {1 Simple S-expression parsing/printing} *)

type t = [`Atom of string | `List of t list]

let equal a b = a = b
let compare a b = Pervasives.compare a b
let hash a = Hashtbl.hash a
let of_int x = `Atom (string_of_int x)
let of_float x = `Atom (string_of_float x)
let of_bool x = `Atom (string_of_bool x)
let atom x = `Atom x
let of_unit = `List []
let of_list l = `List l
let of_rev_list l = `List (List.rev l)
let of_pair (x, y) = `List [x; y]
let of_triple (x, y, z) = `List [x; y; z]
let of_quad (x, y, z, u) = `List [x; y; z; u]
let of_variant name args = `List (`Atom name :: args)
let of_field name t = `List [`Atom name; t]
let of_record l = `List (List.map (fun (n, x) -> of_field n x) l)

(** {6 Traversal of S-exp} *)

module Traverse = struct
  type 'a conv = t -> 'a option

  let return x = Some x
  let ( >|= ) e f = match e with None -> None | Some x -> Some (f x)
  let ( >>= ) e f = match e with None -> None | Some x -> f x

  let map_opt f l =
    let rec recurse acc l =
      match l with
      | [] -> Some (List.rev acc)
      | x :: l' -> (
        match f x with None -> None | Some y -> recurse (y :: acc) l' ) in
    recurse [] l

  let rec _list_any f l =
    match l with
    | [] -> None
    | x :: tl -> (
      match f x with Some _ as res -> res | None -> _list_any f tl )

  let list_any f e = match e with `Atom _ -> None | `List l -> _list_any f l

  let rec _list_all f acc l =
    match l with
    | [] -> List.rev acc
    | x :: tl -> (
      match f x with
      | Some y -> _list_all f (y :: acc) tl
      | None -> _list_all f acc tl )

  let list_all f e = match e with `Atom _ -> [] | `List l -> _list_all f [] l

  let _try_atom e f =
    match e with
    | `List _ -> None
    | `Atom x -> ( try Some (f x) with _ -> None )

  let to_int e = _try_atom e int_of_string
  let to_bool e = _try_atom e bool_of_string
  let to_float e = _try_atom e float_of_string
  let to_string e = _try_atom e (fun x -> x)
  let to_pair e = match e with `List [x; y] -> Some (x, y) | _ -> None

  let to_pair_with f1 f2 e =
    to_pair e
    >>= fun (x, y) -> f1 x >>= fun x -> f2 y >>= fun y -> return (x, y)

  let to_triple e =
    match e with `List [x; y; z] -> Some (x, y, z) | _ -> None

  let to_triple_with f1 f2 f3 e =
    to_triple e
    >>= fun (x, y, z) ->
    f1 x >>= fun x -> f2 y >>= fun y -> f3 z >>= fun z -> return (x, y, z)

  let to_list e = match e with `List l -> Some l | `Atom _ -> None

  let to_list_with f (e : t) =
    match e with `List l -> map_opt f l | `Atom _ -> None

  let rec _get_field name l =
    match l with
    | `List [`Atom n; x] :: _ when name = n -> Some x
    | _ :: tl -> _get_field name tl
    | [] -> None

  let get_field name e =
    match e with `List l -> _get_field name l | `Atom _ -> None

  let field name f e = get_field name e >>= f

  let rec _get_field_list name l =
    match l with
    | `List (`Atom n :: tl) :: _ when name = n -> Some tl
    | _ :: tl -> _get_field_list name tl
    | [] -> None

  let field_list name f e =
    match e with `List l -> _get_field_list name l >>= f | `Atom _ -> None

  let rec _get_variant s args l =
    match l with
    | [] -> None
    | (s', f) :: _ when s = s' -> f args
    | _ :: tl -> _get_variant s args tl

  let get_variant l e =
    match e with
    | `List (`Atom s :: args) -> _get_variant s args l
    | `List _ -> None
    | `Atom s -> _get_variant s [] l

  let get_exn e =
    match e with None -> failwith "CCSexp.Traverse.get_exn" | Some x -> x
end
